### PR TITLE
Worker per connection

### PR DESCRIPTION
**Pull request by pmlopes**

### CODE REVIEW

**Comment by purplefox** [comment]
I'm curious - what's the reasoning behind this?

**Comment by pmlopes** [comment]
@purplefox this is a fix for the issue: https://bugs.eclipse.org/bugs/show_bug.cgi?id=482542

File: src/main/java/io/vertx/ext/jdbc/impl/actions/AbstractJDBCAction.java [hunk 1]
```diff
@@ -26,34 +27,57 @@
...
-  protected AbstractJDBCAction(Vertx vertx, Connection conn) {
+  protected AbstractJDBCAction(Vertx vertx, Connection conn, Context context) {
     this.vertx = vertx;
     this.conn = conn;
+    this.context = context;
   }
 
-  @Override
-  public void handle(Future<T> future) {
+  public void handle(Future<T> future, long timerID) {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/jdbc/impl/actions/AbstractJDBCAction.java]: remove the timeID it was not meant to be here.

File: src/main/java/io/vertx/ext/jdbc/impl/actions/AbstractJDBCAction.java [hunk 2]
```diff
@@ -26,34 +27,57 @@
...
+        vertx.cancelTimer(timerID);
+        future.fail(e);
+      }
     }
   }
 
   public void execute(Handler<AsyncResult<T>> resultHandler) {
-    vertx.executeBlocking(this, resultHandler);
+    Future<T> f = Future.future();
+    long timerID = vertx.setTimer(5000, id -> {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/jdbc/impl/actions/AbstractJDBCAction.java]: remove timer, it was not meant to be here.

**Comment by purplefox** [comment]
Is there an explanation for why the deadlock occurs and a stack trace?

**Comment by vietj** [comment]
the idea is to order the blocking blocks per connection and not for all the connections of the jdbc client

**Comment by vietj** [comment]
it happens because the jdbc executions are on the same worker, so the two locks are executed first, the second lock blocks and the other statements (insert/update and commit / close) are never executed.

**Comment by purplefox** [comment]
I guess I don't quite follow. Any chance you could summarise here with a sequence of actions?

**Comment by vietj** [comment]
yes, imagine you do:

```
void doSomeWork(Connection conn) {
  conn.execute("lock table foo");
  conn.execute("..."); // Do some foo modification
  conn.commit();
}
```

if you have two concurrent execution it will work because one thread will lock the table and then release it and the other can then proceed.

with the jdbc-client this blocks because the sequence can happen:

- lock table foo
- lock table foo
- update foo
- update foo
- commit
- commit

as jdbc-client uses an ordered execute blocking, everything is executed on the same worker. the second lock table is blocking and the update/commit sequence that would release the lock never happens.

**Comment by pmlopes** [comment]
Also to note that this change brings some very impressive performance improvements, here are the techempowert query test before and after the fix:

```
wrk -c 256 -t 32 -d 60 -s ./scripts/pipeline.lua http://localhost:8080/jdbc/queries?queries=5 -- 16

<<< BEFORE >>>

Running 1m test @ http://localhost:8080/jdbc/queries?queries=5
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.00us    0.00us   0.00us    -nan%
    Req/Sec    37.97      9.16    80.00     81.32%
  71006 requests in 1.00m, 19.63MB read
  Socket errors: connect 0, read 0, write 0, timeout 4352
Requests/sec:   1182.22
Transfer/sec:    334.72KB
[plopes@heimdall vertx-benchmark]$ 

<<< AFTER >>>

Running 1m test @ http://localhost:8080/jdbc/queries?queries=5
  32 threads and 256 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   190.55ms  450.67ms   1.85s     0.00%
    Req/Sec    86.80     18.99   160.00     79.77%
  162463 requests in 1.00m, 44.92MB read
  Socket errors: connect 0, read 0, write 0, timeout 512
Requests/sec:   2704.11
Transfer/sec:    765.63KB
```

**Comment by purplefox** [comment]
Ok... but the JDBC execution is ordered for a good reason (otherwise we would just execute executeBlocking with ordered=false).

Without ordering a verticle using the JDBC client can have strange and unexpected results, e.g.:

The following are executed on the same context

1. INSERT INTO foo ("name") values ("tim", "green")
2. SELECT * from foo where name == "tim"

In 2. you would expect it return "tim", "green"

but without ordering the insert might occur after the select which could break a lot of apps.

I suspect the good techempower perf is because dropping the ordering constraint allows more threads from the pool to be used.

Perhaps the jdbc client needs a method:

setOrdered(true/false)

which the user could turn off ordering if it doesn't need it, but I'm not sure turning off ordering is a good idea in general.

**Comment by vietj** [comment]
this branch keeps ordering per SQLConnection, if 1 and 2 are executed with same connection what could go wrong ?

**Comment by purplefox** [comment]
Ok, that makes sense. Thanks for the explanation :)

**Pull request merged**